### PR TITLE
separate simulation time-step and forcing time-step

### PR DESCRIPTION
**Pull request by nmizukami**
Allow user to specify simulation time step regardless of runoff and water-management input time steps.  Use new routine-timeMap_sim_forc to map time indices in runoff or/and water management data for each simulation time step. Allow sub-cycling (use the same forcing at multiple simulation time steps or aggregate forcing if simulation time step is larger than forcing time step).

Clean-ups:

- Removed duplicated routines (get_qix in standalone/model_setup.f90 and match_index in nr_util.f90)
- Removed un-used derived data-type (subbasin_omp_tmp)
- Removed or modify unnecessary components (in input_data, inFileInfo, runoff_data -> input_data and runoff and wm is extend from input_data)

### CODE REVIEW

File: route/build/src/dataTypes.f90 [hunk 1]
```diff
@@ -158,32 +148,36 @@ MODULE dataTypes
...
-   real(dp)                 , allocatable  :: sim(:)        ! flux simulation (HM_HRU) at one time step (size: nSpace(1))
-   real(dp)                 , allocatable  :: sim2D(:,:)    ! flux simulation (x,y) at one time step (size: /nSpace(1),nSpace(2)/)
-   integer(i4b)             , allocatable  :: hru_id(:)     ! id of HM_HRUs or RN_HRUs at which runoff is stored (size: nSpace(1))
-   integer(i4b)             , allocatable  :: hru_ix(:)     ! Index of RN_HRUs associated with river network (used only if HM_HRUs = RN_HRUs)
-   real(dp)                 , allocatable  :: basinRunoff(:)! remapped river network catchment runoff (size: number of nHRU)
-   real(dp)                 , allocatable  :: basinEvapo(:) ! remapped river network catchment runoff (size: number of nHRU)
-   real(dp)                 , allocatable  :: basinPrecip(:)! remapped river network catchment runoff (size: number of nHRU)
+ ! mapping time step between two time series e.g., simulation time step vs runoff time step for one simulation time step
+ type, public :: map_time
+   integer(i4b), allocatable :: iFile(:)
```
Comment (ekluzek) [inline route/build/src/dataTypes.f90:153]: Add some comments about what each of these variables are.

File: route/build/src/dataTypes.f90 [hunk 2]
```diff
@@ -158,32 +148,36 @@ MODULE dataTypes
...
  end type runoff
 
- ! water management data; fluxes to/from reaches or target volume
- type, public :: wm
-   integer(i4b)                            :: nTime           ! number of time steps
-   integer(i4b)                            :: nSpace(1:2)     ! number of spatial dimension, in this case only one dimentonal
-   real(dp)                                :: time            ! time variable at one time step
-   real(dp)                 , allocatable  :: sim(:)          ! user specified flux add/subtract, or volume at one time step (size: nSpace)
-   real(dp)                 , allocatable  :: sim2D(:,:)      ! to provide modularity for reading data
+ type, public, extends(inputData) :: wm
```
Comment (ekluzek) [inline route/build/src/dataTypes.f90:175]: We should add some comments that wm means Water Management

File: route/settings/SAMPLE.control [hunk 3]
```diff
@@ -12,77 +12,93 @@
...
+<route_opt>               1                          ! options for routing schemes (multiple options allowed): 1->IRF 2->KWT 3->KW 4->MC 5->DW
+<doesBasinRoute>          1                          ! basin routing options: 0-> no, 1->IRF, otherwise error
+<restart_write>           last                       ! restart write options: never, daily, monthly, yearly, last, specified
+<restart_date>            yyyy-mm-dd hh:mm:ss        ! restart date.  activated only if <restart_write> is "specified"
+<fname_state_in>          INPUT_RESTART_NC           ! input restart netCDF name. remove or 'coldstart' for run without any particular restart file
+<newFileFrequency>        month                      ! output netcdf frequency: single, daily, monthly, yearly
+<dt_qsim>                 86400                      ! time interval of the forcing [sec] e.g., 86400 sec for daily step
+! lake mode
+<is_lake_sim>             T                          ! switch on (T) or off (F) lake simulation
+<is_flux_wm>              T                          ! switch on (T) or off (F) abstraction from or injection to seg/lakes
```
Comment (ekluzek) [inline route/settings/SAMPLE.control:27]: These three variables aren't very clear as to what their purpose is. Can you help clarify a bit? I'd also give a reminder that "wm" stands for Water Management

**Review by ekluzek — APPROVED** [review]
I have a few suggestions on comments to add. There's a lot of code here, so hard to go through in detail. But, I like changes that I see like in using parameters rather than magic numbers. And the principle of this change is certainly a good one.

I'm approving to make it easier for you to decide how to handle.

File: route/build/src/dataTypes.f90 (hunk 1 above)
Comment (nmizukami) [reply route/build/src/dataTypes.f90]: Added descriptions on `map_time` data type and its components

File: route/build/src/dataTypes.f90 (hunk 2 above)
Comment (nmizukami) [reply route/build/src/dataTypes.f90]: Added a comment "water-management" next to `wm`

File: route/settings/SAMPLE.control (hunk 3 above)
Comment (nmizukami) [reply route/settings/SAMPLE.control:27]: modified comment "!lake mode" to "!lake and water-management model"

**Comment by nmizukami** [comment]
Thank you Erik for comments!

**Pull request merged**